### PR TITLE
Adding S3 support for HTTP domain validation

### LINKED ISSUE #1969: Add S3 as HTTP provider
### Welcome

- [X] Yes, I've searched similar issues on GitHub and didn't find any.

### How do you use lego?

Through Terraform ACME provider

### Detailed Description

I would like to be able to use s3 to host the Domain Validation Token

The provider would only be responsible for uploading the token to the .well-known/acme-challenge/<TOKEN> path within the already existing s3 bucket.

It will be assumed that the domain name(s) are already pointing to the s3 bucket and it'll be available via http or https

https://github.com/vancluever/terraform-provider-acme/issues/327

### PR DESCRIPTION
**Pull request by kingcdavid**
PR to allow upload token to S3 for HTTP domain Validation

Fixes #1969
Closes #334

### CODE REVIEW

**Comment by ldez** [comment]
I think you forgot to add the provider code itself.

**Comment by kingcdavid** [comment]
That's embarrassing, forgot the git add 

updated now

**Comment by ldez** [comment]
I follow your PR and migrate route53 and lightsail to the SDK v2: #1973

**Comment by ldez** [comment]
I will ask you to provide a command example and the output of this command. (with redacted information :wink:)

**Comment by ldez** [comment]
@beornf could you try this PR?

**Comment by kingcdavid** [comment]
> I will ask you to provide a command example and the output of this command. (with redacted information wink)

`dist/lego -k rsa2048 --domains "<redacted>" --server https://emea.acme.atlas.globalsign.com/directory --kid <redacted> --eab --hmac <redacted>  --http.s3-bucket daveacmetest --http --email <redacted> run 
2023/07/25 13:04:34 [INFO] [<redacted>] acme: Obtaining bundled SAN certificate
2023/07/25 13:04:35 [INFO] [<redacted>] AuthURL: https://emea.acme.atlas.globalsign.com/authz/<redacted>
2023/07/25 13:04:35 [INFO] [<redacted>] acme: use http-01 solver
2023/07/25 13:04:35 [INFO] [<redacted>] acme: Trying to solve HTTP-01
2023/07/25 13:04:41 [INFO] [<redacted>] The server validated our request
2023/07/25 13:04:41 [INFO] [<redacted>] acme: Validations succeeded; requesting certificates
2023/07/25 13:04:42 [INFO] Wait for certificate [timeout: 30s, interval: 500ms]
2023/07/25 13:04:42 [INFO] [<redacted>] Server responded with a certificate.
`

**Review by beornf — CHANGES_REQUESTED** [review]
Missed testing the functionality `provider.cleanUp` in the s3 test method `TestLiveS3ProviderPresent`.

File: providers/http/s3/s3_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	provider, err := NewHTTPProvider(s3Bucket)
+	require.NoError(t, err)
+
+	err = provider.Present(domain, token, keyAuth)
+	require.NoError(t, err)
+
+	// Need to wait a little bit before checking website
+	time.Sleep(1 * time.Second)
+
+	chlgPath := fmt.Sprintf("http://%s.s3-website.%s.amazonaws.com%s",
```
[the remark this answers is not in the document]
Comment (kingcdavid) [reply providers/http/s3/s3_test.go]: Have updated

File: providers/http/s3/s3_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+	// Need to wait a little bit before checking website aghain
+	time.Sleep(1 * time.Second)
+
+	cleanupResp, err := http.Get(chlgPath)
+	require.NoError(t, err)
+
+	defer func() { _ = resp.Body.Close() }()
+
+	assert.Equal(t, cleanupResp.StatusCode, 404)
```
Comment (beornf) [inline providers/http/s3/s3_test.go:83]: I get a 403 here. Did you test this on the new path?

Comment (kingcdavid) [reply providers/http/s3/s3_test.go:83]: I did, it depends on if you have the public list permission set on your s3 bucket
```console
curl -s -o /dev/null -w "%{http_code}" http://<redacted>.s3.eu-west-2.amazonaws.com/.well-known/acme-challenge/foo
404
```
Happy to change it to 403 if we think it's a more common permission, I've confirmed that is the case if i remove the Public List permission

Comment (kingcdavid) [reply providers/http/s3/s3_test.go:83]: Actually, i think 403 is probably more common, so have updated

**Comment by ldez** [comment]
can you add a small readme inside the package to explain how to set it up (permissions, expectations, ...)?

**Review by ldez — APPROVED** [review]
LGTM

**Pull request merged**